File: String Algorithms/Questions/first_non_repeating_character.py
def get_character_index(s):
    d = {}  # will hold values in the format[fist_index, count]

    for i in range(len(s)):
        if d.get(s[i]):
            d[s[i]][1] += 1
        else:
            d[s[i]] = [i, 1]

    min_index = len(s) + 1  # will hold the index of the first character with frequency 1
    
    for i in d:
        if d[i][1] == 1 and min_index > d[i][0]:
            min_index = d[i][0]

    return min_index


if __name__ == "__main__":
    string = input("Enter string\n")
    index = get_character_index(string)

    try:
        print("First Non-repeating character is {} at index {}".format(string[index], index))
    except:
        print("No Non-repeating character")

File: String Algorithms/Questions/FirstNonRepeatingCharacter.java
import java.util.*;
import java.text.*;

class NonRepeating
{
    public static char getFirstNonRepeatingCharacter(String s) {
        if (s.length() == 1)
            return s.charAt(0);

        HashMap<Character, Integer> counter = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i) ;
            counter.put(c, counter.getOrDefault(c, 0) + 1);
        }

        for (int i = 0; i < s.length(); i++)
            if (counter.get(s.charAt(i)) == 1)
                return s.charAt(i);

        return '_';
    }


    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter a string:");
        String string = scanner.nextLine();
        System.out.println(MessageFormat.format("First Non Repeating Character: {0}", getFirstNonRepeatingCharacter(string)));
    }
}